pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/TransactionTest.java

```diff
@@ -696,4 +696,33 @@ public void testEndTxnWhenCommittingOrAborting() throws Exception {
         commitTxn.commit();
     }
 
+    @Test
+    public void testCommittedWhenHasTxnOpFailed() throws Exception {
+        Transaction transaction = pulsarClient
+                .newTransaction()
+                .withTransactionTimeout(5, TimeUnit.SECONDS)
+                .build()
+                .get();
+
+        Producer<String> producer = pulsarClient
+                .newProducer(Schema.STRING)
+                .topic(NAMESPACE1 + "/test")
+                .sendTimeout(0, TimeUnit.SECONDS)
+                .create();
+        for (PulsarService pulsarService : getPulsarServiceList()) {
+            Field field = ManagedLedgerImpl.class.getDeclaredField("state");
+            field.setAccessible(true);
+            field.set(((MLTransactionMetadataStore) pulsarService.getTransactionMetadataStoreService().getStores()
+                    .get(TransactionCoordinatorID.get(0))).getManagedLedger(), ManagedLedgerImpl.State.Closed);
+        }
+        producer.newMessage(transaction)
+                .value("test")
+                .sendAsync();
```

Comment:
Here we can have a race.
Because we are not waiting for this CompletableFuture to finish, so we could succeed in committing the transaction is the send op has not been processed but the server

Reply:
When we commit or abort a transaction, it will wait for all CompletableFutures to finish.